src/analysis/safety_assertion_mode.rs

```diff
@@ -0,0 +1,36 @@
+use analysis::parameter::Parameter;
+use env::Env;
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum SafetyAssertionMode {
+    None,
+    Skip,
+    InMainThread,
+}
+
+impl Default for SafetyAssertionMode {
+    fn default() -> SafetyAssertionMode {
+        SafetyAssertionMode::None
+    }
+}
+
+impl SafetyAssertionMode {
+    pub fn of(env: &Env, params: &[Parameter]) -> SafetyAssertionMode {
+        use self::SafetyAssertionMode::*;
+        use library::Type::*;
+        if !env.config.generate_safety_asserts {
+            return None;
+        }
+        if params.len() > 0 && params[0].instance_parameter {
+            return None;
+        }
+        for par in params {
+            match *env.library.type_(par.typ) {
+                Class(..) => return Skip,
```

Comment:
Also `Interface`.

This gets into somewhat difficult area though: `Option<T>` or `&[T]` might allow calling the function without a previously created object (e.g. `Layout::new`).
It might be better to merge with just the `instance_parameter` case handled and then add improvements if any.